### PR TITLE
closes #6384: Handle dot-style wildcard domains like star-style domains

### LINKED ISSUE #6384: [Feature request] nginx plugin doesn't auto detect wildcard domains
## My operating system is (include version):
Ubuntu 18.04.1

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):

apt initially, then pip installed all certbot packages to see if issue was resolved

## I ran this command and it produced this output:

    sudo certbot --authenticator=dns-route53 --installer=nginx

    No names were found in your configuration files. Please enter in your domain

## Certbot's behavior differed from what I expected because:

I have a single _index_ nginx conf containing a wildcard `server_name` which routes to all other 'location' confs based on path

Why can't the nginx plugin just detect that its a wildcard cert?

```
server {
    listen 80;
    server_name .example.com;

    root /var/www;
    index index.html;

    include conf.d/*.location;
}
```

### PR DESCRIPTION
**Pull request by ciarancourtney**
* e.g. mutate .example.com into *.example.com as early as possible

Not sure if this is the correct approach, but seems reasonable 

closes #6384

### CODE REVIEW

**Comment by bmw** [comment]
@ohemorange, it looks like you were talking to @ciarancourtney about this issue. Are you able to this PR?

**Comment by ciarancourtney** [comment]
Hi @ohemorange just pushed 983653ffe84b0f9ab25dc47f8810651838e75acb, is that what you had in mind? If so, I think there was an implicit bug in some tests relating to the example.com case

**Comment by ohemorange** [comment]
Changing it there would actually change how the configurator sees the server names in various different places. This leads to some different behavior like copying over server names incorrectly. It might be fine, but it's probably best to keep the changes as small as possible to fix this one particular thing, and we also already handle this case specifically in other places when we need to. I recommend modifying the domains in the `get_all_names` methods of `certbot-nginx/certbot_nginx/configurator.py`.

**Comment by ciarancourtney** [comment]
Hi @ohemorange  I pushed a change in approach, does this seem better? One travis test is failing, doesn't make sense to me

```
            for server_name in vhost.names:
>               if util.is_wildcard_dot_domain(server_name):
E               AttributeError: 'module' object has no attribute 'is_wildcard_dot_domain'
```

**Comment by ohemorange** [comment]
To fix that test, bump the version of certbot required by certbot-nginx to 0.29.0; an similar example of how to do that is here (no need to change `local-oldest-requirements.txt` or `setup.py`, just the first two files, but for `certbot`): https://github.com/certbot/certbot/pull/6250/files

**Comment by ohemorange** [comment]
But actually, since the wildcard dot domain is Nginx-specific, there probably don't have to be any code changes in `certbot` at all; if you move the util function to `certbot-nginx` and do what you're doing in `_get_all_names` (which looks great!), there probably don't need to be any other changes in `certbot`.

**Comment by ciarancourtney** [comment]
Hi @ohemorange just rebased but 2 builds still failing for some reason, can't figure out why, maybe you know?

**Comment by ohemorange** [comment]
Certbot's latest version is no longer `0.29.0`; it should now be `0.31.0`, which you can find here: https://github.com/certbot/certbot/releases

That should account for at least the Apache test failures.

**Comment by ohemorange** [comment]
Also, we squash commits, so it's actually a little cleaner to follow the history if you merge in master rather than rebasing.

File: certbot-nginx/local-oldest-requirements.txt [hunk 1]
```diff
@@ -1,2 +1,2 @@
 acme[dev]==0.26.0
-certbot[dev]==0.22.0
+certbot[dev]==0.31.0
```
Comment (ohemorange) [inline certbot-nginx/local-oldest-requirements.txt]:
```suggestion
-e .[dev]
```

File: certbot-nginx/setup.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 # acme/certbot version.
 install_requires = [
     'acme>=0.26.0',
-    'certbot>=0.22.0',
+    'certbot>=0.31.0',
```
Comment (ohemorange) [inline certbot-nginx/setup.py]:
```suggestion
    'certbot>=0.32.0.dev0',
```

File: certbot-nginx/local-oldest-requirements.txt (hunk 1 above)
Comment (ciarancourtney) [reply certbot-nginx/local-oldest-requirements.txt]: Hi @ohemorange Tried this but it still fails. You mentioned apache, should the changes be made there also or instead? Feel free to mod this PR, I have no clue of the testing deps of certbot.

Comment (ohemorange) [reply certbot-nginx/local-oldest-requirements.txt]: Yes, you're right, Apache needs the change as well. I can't push to your branch, the change is the same as in `certbot-nginx/setup.py`.

**Comment by ohemorange** [comment]
The Nginx compatibility failure looks like it's a real failure due to a change in behavior; you'll probably want to look at `_get_names` in `certbot-compatibility-test/configurators/nginx/common.py`.

**Comment by ohemorange** [comment]
Hey @ciarancourtney, what is your current blocker? Could you use more clarity about the Nginx compatibility test failures?

**Comment by ciarancourtney** [comment]
> Hey @ciarancourtney, what is your current blocker? Could you use more clarity about the Nginx compatibility test failures?

I tried that failing tox test locally and it passed, the travis result doesn't say what failed https://travis-ci.com/certbot/certbot/jobs/179620921

**Comment by ohemorange** [comment]
It doesn't pass locally for me. The issue probably has to do with these configs:

```
nginx: [warn] conflicting server name "random1413.example.org" on 0.0.0.0:443, ignored
nginx: [warn] conflicting server name "www.random1413.example.org" on 0.0.0.0:443, ignored
nginx: [warn] conflicting server name "random25266.example.org" on 0.0.0.0:443, ignored
nginx: [warn] conflicting server name "www.random25266.example.org" on 0.0.0.0:443, ignored
nginx: [warn] conflicting server name "random28524.example.org" on 0.0.0.0:443, ignored
nginx: [warn] conflicting server name "www.random28524.example.org" on 0.0.0.0:443, ignored
nginx: [warn] conflicting server name "random27579.example.org" on 0.0.0.0:443, ignored
nginx: [warn] conflicting server name "www.random27579.example.org" on 0.0.0.0:443, ignored
nginx: [warn] conflicting server name "random25863.example.org" on 0.0.0.0:443, ignored
nginx: [warn] could not build optimal server_names_hash, you should increase either server_names_hash_max_size: 512 or server_names_hash_bucket_size: 64; ignoring server_names_hash_bucket_size
nginx: [warn] conflicting server name "random28524.example.org" on 0.0.0.0:80, ignored
nginx: [warn] conflicting server name "www.random28524.example.org" on 0.0.0.0:80, ignored
nginx: [warn] conflicting server name "" on 0.0.0.0:80, ignored
nginx: [warn] conflicting server name "www.random25863.example.org" on 0.0.0.0:80, ignored
nginx: [warn] conflicting server name "random25863.example.org" on 0.0.0.0:80, ignored
nginx: [warn] could not build optimal server_names_hash, you should increase either server_names_hash_max_size: 512 or server_names_hash_bucket_size: 64; ignoring server_names_hash_bucket_size
```

**Comment by ohemorange** [comment]
Hey @ciarancourtney, how's investigation on this going?

**Comment by ciarancourtney** [comment]
Hi @ohemorange I rebased and it seems some failures related to certbox package versions were fixed upstream, by `nginx_compat` seems to have one failure I can repro locally with tox, but I can't see how you could identify where the failure is coming from.

See [here](http://www.mergely.com/xCFZEC6S/) for a diff between a good travis run (left) and this PR (right) and search `get_all_names test failed for config `

**Comment by ohemorange** [comment]
Hey @ciarancourtney, looks like there's still some test failures going on here. Unfortunately we don't have the bandwidth to handle the debugging on our end, so I'm going to have to close the issue. If you do end up being able to resolve them please do let us know.

**Pull request closed without merging**